### ISSUE TITLE
Allow ordering of columns in Data Model

### ISSUE THREAD

**Issue by tomsej** [metabase/metabase#3720]
Currently it is not possible to change ordering of columns in Data Model tab. It seems that columns are ordered by database column name. If the business name is different to database column name, it is hard to find requested field.

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by HelmiRifai** [comment]
+1

**Comment by LukeAbell** [comment]
+1

**Comment by philip-n** [comment]
Just started to create a new issue, but found this one just in time. Since I already wrote "my" issue, I'll just paste it here anyway ;)

---

For my use case, **preserving and/or manually arranging the order of the columns** would be very useful. This could be done either during sync (i.e. to preserve the original order of the columns) or by having a possibility to manually arrange them in the Admin > Data Model view. 

Example:
- A table has four columns named "First", "Second", "Third", "Fourth".
- Current situation: They will be listed as "First", "Fourth", "Second", "Third". _This is true for both views in Admin > Data Model (show original schema turned on / off) as well as e.g. the "raw data" output for questions._
- Proposal / wish: Add the possiblilty to define a custom ordering, to make the columns show up in their original order ("First", "Second", "Third", "Fourth") and/or in a custom order ("Fourth", "Third", "Second", "First")

Currently, there is the possibility to "customize this table" when adding a new question, where the fields can be (de)selected and their order can be arranged. However, this reordering needs to be done for each question --> having a custom order in the data model would be a much more elegant solution.

**Comment by MarkRx** [comment]
The column ordering for Oracle views is resulting in some random order which is quite annoying.

**Comment by mvrabel** [comment]
@salsakran could you add priority to this please ? 
**This issue would save a lot o time for us**
Each time we want to create a new question we have tu manually rearange the columns for each question. 

This is especialy **tedious** when you have a table with **~40 columns.**
If we could add a default ordering **OR** if the qustion would show columns in order as in the database table that would be great.
Thanks,
Martin

**Comment by mazameli** [comment]
Thanks for the feedback; definitely makes sense. One workaround for now, if you're doing any querying with the graphical custom query builder, would be to save a question that just returns all the raw data in a particular table, but with the columns reordered the way you'd like. Then you can use that saved question as a pseudo data table for new custom questions. More on these "nested" questions here: https://www.metabase.com/blog/Nested-Queries/

**Comment by mvrabel** [comment]
@mazameli Thanks for hte quick reply. I know about this workaround.
**The problem** is that it's not possible to use segments if you do that :(

**Comment by fredericocarneiro** [comment]
+1

**Comment by vvaezian** [comment]
That would be great if in custom questions, columns are ordered as in the database by default (or have the option to enforce this in the data model).
I see the power of Metabase as providing non-technical users a UI to interact with the data (in addition to dashboarding of course). The default alphabetical sorting of columns in custom questions makes this a bit hard.
Thanks.

**Comment by Crusader4Christ** [comment]
@mazameli Second problem is - filter id widgets not working (Not filter by name, not show name by id)

**Comment by Nico-P-FR** [comment]
Same here. It would be very useful to let us reorganize fields order or at least choose between automatic alphabetical order and database original order.

**Comment by ACPK** [comment]
@tomsej  - Are users still not able to order tables without using custom SQL?

**Comment by rarup1** [comment]
This is a very important feature for our usage too. It is fine to use the work around mentioned above but when you ask someone to use your data structure from the raw tables in the DW (which is far more complete) they will have to go through the pain of unticking and organising all over again. 

Is there anyone that is looking into adding a setting within the data model that allows either alphabetical ordering as it is at the moment or db ordering as collected from the information schema?

Would try to dive into the code myself but I am sure someone who knows the project could do it in a heartbeat...

**Comment by antoine-lizee** [comment]
In our case:
- We don't need to be able to decide the order of individual fields in metabase, and I don't think most people here do.
- We need to have the database order be kept because currently, the exploration of data is close to impossible as soon as there are more than 10 fields. The reshuffle of the fields that Metabase does completely undermines our current push for promoting the query builder to interact with the data we prepared. (we get back to SQL each time just for this reason... `SELECT * FROM xxx` doesn't change order!)
- I don't see a viable use case for reshuffling the column order, so I would just remove this reordering. But if you are afraid of back-compatibility it makes sense to add a general toggle between alphabetical / database order.

🙏 Please 🙏 don't let the potential of building a complete ordering feature maybe prevent you from shipping the simple fix!

**Comment by lucascorreaaa** [comment]
This feature is theoretically simple and it would be really handy. Hope someone can fix it soon.

**Comment by yskolnick** [comment]
Is this feature being worked on?

**Comment by yskolnick** [comment]
I am willing to 

> In our case:
> 
> * We don't need to be able to decide the order of individual fields in metabase, and I don't think most people here do.
> * We need to have the database order be kept because currently, the exploration of data is close to impossible as soon as there are more than 10 fields. The reshuffle of the fields that Metabase does completely undermines our current push for promoting the query builder to interact with the data we prepared. (we get back to SQL each time just for this reason... `SELECT * FROM xxx` doesn't change order!)
> * I don't see a viable use case for reshuffling the column order, so I would just remove this reordering. But if you are afraid of back-compatibility it makes sense to add a general toggle between alphabetical / database order.
> 
> 🙏 Please don't let the potential of building a complete ordering feature prevent you from shipping the simple fix!

I am looking through the code to see where the reordering takes place and I am having a hard time locating it.   Anyone have any idea?

**Comment by sajjadkhanGit** [comment]
This is a big issue, Did someone find any solution to this?

**Comment by sajjadkhanGit** [comment]
Metabase should have picked the order of the columns as it is coming from the Source DB table.
Imagin 50 columns of a table and re-arranging each time we click on the table is much hard-work.

**Comment by Nico-P-FR** [comment]
No, still have this problem and it is very painfull for us too. We have table with more than 100 columns... If the feature "allow ordering" is too long due to UI, a MVP could be to switch from automatic alphabetical ordering to keeping the DB Source order.
Even the Metabase field metabase_field.position is already there, it is just not used...

**Comment by mazameli** [comment]
I agree that the incremental step we should seriously consider taking on this is to **preserve the order of columns as they are in the database.**

This feels similar to the "friendly field names" behavior that we undid — we're currently trying to be overly clever and in the process are undermining what people actually want and expect.

**Comment by arcreative** [comment]
I think it's great that the columns have the ability to be reordered and removed in the UI, but I think the original order should be default, with an option in the modal to order alphabetically or by friendly name.  Agree with all the above, and propose something like the following (after the initial viz creation):

![reorder](https://user-images.githubusercontent.com/1896026/73385642-5952c600-4282-11ea-90ef-79430f7c89e6.png)

So I would suggest keeping the ability to select/deselect rows (alphabetical if you must) in the creation UI, but allow us to reorder them in the question itself with something like the above screenshot.

**Comment by sajjadkhanGit** [comment]
Hi Guys! I've done some workaround in the source DB table and successfully managed to set the desired order of columns.  Here is solution for you--

1) Use select statement and doing so set the alias of each column as follows 
    [a% columnname1]
    [b% columnname2]
    [c% columnname3]
    [d% columnname4]
    .
    .
    So on
2) In the Metabase table, you will see the same order as above but with % sign

3) If you want to get rid of this format and want to show without a%, b% in Metabase then you need to do following step in Metabase DB in my case it is MySQL. 
So go to the Metabase table metabase_field and update the display_name field as follows:
**update metabase.metabase_field m set display_name=SUBSTRING_INDEX(display_name, '% ', -1)**

Now your columns will be shown in order.

Thanks

**Comment by vvaezian** [comment]
I've written a Python wrapper that (among other things) makes it possible to create cards with the desired column order:
```python
from metabase_api import Metabase_API

mb = Metabase_API('https://...', 'username', 'password')
mb.create_card(card_name='test_card', table_name='mySourceTable')
``` 
The `column_order` parameter specifies how to handle the order of columns:
- 'db_table_order' (default): The column order used in the DB table
- 'alphabetical'
- You can also provide a list of column names to specify a custom order (also useful if you don't want all of the columns be used in the card).

More details [here](https://github.com/vvaezian/metabase_api_python).

**Comment by mattarderne** [comment]
Related - is there a bug for dragging a field "selects" other fields as you move fields around? Combining that with this issue has caused me much frustration.

<img width="178" alt="image" src="https://user-images.githubusercontent.com/23472325/77664519-47398f00-6f76-11ea-9dff-1455c696709f.png">

Firefox, latest Metabase

**Comment by camsaul** [comment]
Notes from our meeting about this today:

- There are either 3 or 4 sort styles (TBD): alphabetical, database order, custom, and "smart sort" (sort ID/names first, then sort by custom order, then alphabetically -- this is the current behavior)
- We currently record Field `position` (custom sort order) -- in an older version of Metabase you could change this in the UI. Currently, it's `0` for all Fields. 
- We should add a `db_position` column to complement the `position` and set the value of this as part of sync
- We should add a "sort style" column to Table -- when new Fields are added, the Table should be re-sorted according to these rules
  -  For *existing* Tables, it's TBD whether what we should set as their default sort style -- and whether we want to resort everything on the next sync if we change it to something other than smart sort
  - We should be able to change this via the API -- and when the value changes, re-sort all of the Fields in the Table (i.e., update their `position`) according to the rule.
  - Changing the `position` of a Field manually should change the Table sort style back to "custom"
- I think sync should also update the `position` of each column according to the sort style for the Table. That way we don't need to duplicate the sort logic in other places that need the sorted columns -- the QP and API endpoints
-  We should remove the "smart sort" logic from the QP middleware that adds the implicit `:fields` clause and have it use `position` directly instead, assuming the `position`s are sorted according to the Table sort style
-  API endpoint that powers data model/etc. (?) should sort Fields according to sort order rather than alphabetically

**Comment by camsaul** [comment]
I think FE work here is just limited to making the admin data model support drag-and-drop reordering for Fields, and adding the drop-down for Table "sort style"

**Comment by sbelak** [comment]
Should the setting be per-table, per-db, or per-instance?

**Comment by ghomem** [comment]
I am seeing the drag and drop problem mentioned by @mattarderne . Firefox 84.0.2. Metabase v0.37.5-SNAPSHOT using this:

https://github.com/metabase/metabase/pull/13081/files

**Comment by ghomem** [comment]
Drag and drop works on some columns but does not work on others (column goes back to the initial place). Tried Firefox and Chrome.

**Comment by flamber** [comment]
@ghomem Please try an official release - I don't know what state that PR is in. And if you are still seeing it on a release, then please open a new issue after you have searched for existing issues - I have a feeling that you're seeing #13024

**Comment by ghomem** [comment]
Thank you for your support again @flamber . We will check. I thought that PR would be new enough to already have this, but maybe not.

**Comment by eozturkTF** [comment]
Hello
Very interested by this feature too.

**Comment by flamber** [comment]
@eozturkTF It has been available since 0.36.0: https://www.metabase.com/blog/v36-release-announcement/index.html

**Comment by eozturkTF** [comment]
thank you
